contracts/Vault.sol

```diff
@@ -674,9 +678,39 @@ contract Vault is
             );
         }
 
+        _rebalanceBeforeTransfer(amount);
+
         underlying.safeTransfer(_to, amount);
     }
 
+    /**
+     * Rebalances the vault's funds to cover the transfer of funds from the vault
+     * by disinvesting from the strategy. After the rebalance the vault is left
+     * with a set percentage (100% - invest%) of the total underlying as reserves.
+     *
+     * @notice this will have effect only for sync strategies.
+     *
+     * @param _amount Funds to be transferred from the vault.
+     */
+    function _rebalanceBeforeTransfer(uint256 _amount) internal {
+        uint256 vaultBalance = underlying.balanceOf(address(this));
+
+        if (_amount <= vaultBalance) return;
+        if (!strategy.isSync()) revert VaultNotEnoughFunds();
+
+        uint256 expectedReserves = (totalUnderlying() - _amount).pctOf(
+            10000 - investPct
+        );
+
+        // we want to withdraw the from the strategy only what is needed
+        // to cover the transfer and leave the vault with the expected reserves
+        uint256 needed = _amount + expectedReserves - vaultBalance;
```

Comment:
another thought that just came to mind. it's arguably very inefficient to always just calculate `needed` and do a withdrawal from the vault. what if `needed = 1`? What if 10 users withdraw from the vault in quick succession? We're incuring on them the gas cost for 10 different rebalances, when we could just have some threshold

e.g.:

```
uint256 desired = _amount + expectedReserves - vaultBalance;
uint256 needed = desired * 90 / 100; // let's say we have a 10% margin of error threshold

// withdraw the original `desired` amount, but only if we don't have enough to cover `needed`, or to cover the withdrawal itself
if (_amount < vaultBalance || needed < vaultBalance) {
  strategy.withdrawToVault(desired);
}
```

@gabrielpoca thoughts?

Reply:
I may be confusing things, but I don't think that scenario exists. The "needed" will always be more than 5% of the total underlying (100% - investment%), because otherwise, this function would not have been called, right? In that scenario where those 10 users withdraw, the one that tries to take more than what's available in the vault would be the only one rebalancing, unless those are big whale deposits that take more than 5% to withdraw. I guess the issue you're pointing out is that one person would be paying for everyone else? Do we pay fees depending on the amount we withdraw?

Reply:
Ah, sorry. I forgot when doing this second review that there's a `if (_amount <= vaultBalance) return;` above